Source/HTTPHeaders.swift

```diff
@@ -24,51 +24,215 @@
 
 import Foundation
 
-public typealias HTTPHeaders = [String: String]
+public struct HTTPHeaders {
+    private var headers = [HTTPHeader]()
+    
+    public init() { }
+    
+    public init(_ headers: [HTTPHeader]) {
+        self.headers = headers
+    }
+    
+    public init(_ dictionary: [String: String]) {
+        self.init()
+        
+        headers = dictionary.map { HTTPHeader(name: $0.key, value: $0.value) }
+    }
+    
+    public mutating func update(name: String, value: String) {
```

Comment:
Currently, `lowercased()` is being called twice for each comparison. It would make more sense to inline the comparison function and also make `update(name:value:)` the convenience function instead of `update(_:)`:

```swift
    mutating func update(_ header: HTTPHeader) {
        let lowercasedName = header.name.lowercased()
        guard let index = headers.firstIndex(where: { $0.name.lowercased() == lowercasedName } ) else {
            headers.append(header)
            return
        }
        headers.replaceSubrange(index...index, with: [header])
    }
    mutating func update(name: String, value: String) {
        update(HTTPHeader(name: name, value: value))
    }
```

Reply:
I've switched the convenience but I made the `lowercased` change in my `index(of:)` extension.